pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+    close (toWalletPipe[1]);
+
+    //Change to the user in case we are not it yet
+    if (drop_privileges(userInfo) < 0) {
+        syslog(LOG_ERR, "%s: could not set gid/uid/euid/egit for kwalletd", logPrefix);
+        goto cleanup;
+    }
+
+    int envSocket;
+    if ((envSocket = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
+        syslog(LOG_ERR, "%s: couldn't create socket", logPrefix);
+        goto cleanup;
+    }
+
+    struct sockaddr_un local = {};
+    local.sun_family = AF_UNIX;
+
+    if (strlen(fullSocket) > sizeof(local.sun_path)) {
+        syslog(LOG_ERR, "%s: socket path %s too long to open",
+                   logPrefix, fullSocket);
+        free(fullSocket);
+        goto cleanup;
+    }
+    strcpy(local.sun_path, fullSocket);
+    unlink(local.sun_path);//Just in case it exists from a previous login
+
+    syslog(LOG_DEBUG, "%s: final socket path: %s", logPrefix, local.sun_path);
+
+    if (bind(envSocket, (struct sockaddr *)&local, sizeof(local)) == -1) {
+        syslog(LOG_INFO, "%s-kwalletd: Couldn't bind to local file\n", logPrefix);
+        goto cleanup;
+    }
+
+    if (listen(envSocket, 5) == -1) {
+        syslog(LOG_INFO, "%s-kwalletd: Couldn't listen in socket: %d-%s\n", logPrefix, errno, strerror(errno));
+        goto cleanup;
+    }
+    //finally close stderr
+    close(2);
+
+    // Fork twice to daemonize kwallet
+    setsid();
+    pid_t pid = fork();
+    if (pid != 0) {
+        if (pid == -1) {
+            exit(EXIT_FAILURE);
+        } else {
+            exit(0);
+        }
+    }
+
+    //TODO use a pam argument for full path kwalletd
+    char pipeInt[4];
+    sprintf(pipeInt, "%d", toWalletPipe[0]);
+    char sockIn[4];
+    sprintf(sockIn, "%d", envSocket);
+
+    char *args[] = {strdup(kwalletd), "--pam-login", pipeInt, sockIn, NULL, NULL};
+    execve(args[0], args, pam_getenvlist(pamh));
+    syslog(LOG_ERR, "%s: could not execute kwalletd from %s", logPrefix, kwalletd);
+
+cleanup:
+    exit(EXIT_FAILURE);
+}
+
+static int better_write(int fd, const char *buffer, int len)
+{
+    size_t writtenBytes = 0;
+    while(writtenBytes < len) {
+        ssize_t result = write(fd, buffer + writtenBytes, len - writtenBytes);
+        if (result < 0) {
+            if (errno != EAGAIN && errno != EINTR) {
+                return -1;
+            }
+        }
+        writtenBytes += result;
+    }
+
+    return writtenBytes;
+}
+
+static void start_kwallet(pam_handle_t *pamh, struct passwd *userInfo, const char *kwalletKey)
+{
+    //Just in case we get broken pipe, do not break the pam process..
+    struct sigaction sigPipe, oldSigPipe;
+    memset (&sigPipe, 0, sizeof (sigPipe));
+    memset (&oldSigPipe, 0, sizeof (oldSigPipe));
+    sigPipe.sa_handler = SIG_IGN;
+    sigaction (SIGPIPE, &sigPipe, &oldSigPipe);
+
+    int toWalletPipe[2] = { -1, -1};
+    if (pipe(toWalletPipe) < 0) {
+        pam_syslog(pamh, LOG_ERR, "%s: Couldn't create pipes", logPrefix);
+    }
```

Comment:
why should the PAM module continue running if this pair of pipes could not be created? seems like we're on a road to doom in this case.